src/main/java/core/basesyntax/ArrayList.java

```
@@ -1,48 +1,185 @@
 
     @Override
     public void addAll(List<T> list) {
+        int newSize = size + list.size();
+        if (newSize < array.length) {
+            copyArray(list);
+        } else {
+            checkIfArrayShouldGrow(list);
+            copyArray(list);
+        }
 
+        size += list.size();
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T get(int index) throws ArrayListIndexOutOfBoundsException {
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        T result = (T) array[index];
+        return result;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public void set(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexNegative(index);
+        checkIfPositionExistent(index);
+        Object[] result = Arrays.copyOf(array, array.length - 1);
+        T removedItem = null;
+        removedItem = (T) array[index];
+        System.arraycopy(array, index + 1, result, index, array.length - index - 1);
+        array = result;
+        size--;
+        return removedItem;
     }
 
     @Override
-    public T remove(T element) {
+    public T remove(T element) throws NoSuchElementException {
+        boolean flag = false;
+
+        for (int i = 0; i < size; i++) {
+            if (array[i] == null) {
+                if (element == null) {
+                    remove(getIndex(element));
+                    flag = true;
+                    return element;
+                }
+                continue;
+            }
+            if (array[i] == element || array[i].equals(element)) {
+                remove(getIndex(element));
+                flag = true;
+                return element;
+            }
+
+        }
+        if (flag == false) {
```

This condition check is unnecessary. You can throw the exception directly after the loop if the element was not found and removed.